SmolVLM2

orrzohar opened the pull request:
# What does this PR do?


SmolVLM2 support

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [yes] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [no] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

orrzohar committed:
smolvlm init

orrzohar committed:
updates

orrzohar committed:
fixing bugs

orrzohar committed:
minimal run, no checks

orrzohar committed:
minimal run, no checks

orrzohar committed:
passing first check + adding url support

orrzohar committed:
updating video dataloading logic

orrzohar committed:
fixing image logic

merveenoyan commented:
cc @ArthurZucker  fyi this is needed for release otherwise it will be too much work on user's side 🥲

molbap commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,480 @@
+            prompt_images.append(images)
+        return prompt_images
+
+    def __call__(
+        self,
+        images: Union[ImageInput, List[ImageInput], List[List[ImageInput]]] = None,
+        video: Union[str, List[ImageInput], List[List[ImageInput]]] = None,
+        video_sampling_fps: int = None,
+        max_frames: int = None,
+        text: Union[TextInput, "PreTokenizedInput", List[TextInput], List["PreTokenizedInput"]] = None,
+        messages: Optional[List[Dict[str, Any]]] = None,
+        add_generation_prompt: bool = False,
+        audio=None,
+        image_seq_len: Optional[int] = None,
+        **kwargs: Unpack[SmolVLMProcessorKwargs],
```
Some API issues here: `__cal__` processor method should be as close as possible to the preexisting methods. The `SmolVLMProcessingKwargs` above hold the types and default values of your passed kwargs, and should be placed there. In particular you will want to add a `...SmolVLMVideosKwargs` class, that can contain `max_frames` and `video_sampling_fps` for instance. Take a look at Qwen 2.5 VL, recently merged with video processing, to get a sense of the API. 

All processors should be as close as possible to
```python
    def __call__(
        self,
        images: Union[ImageInput, List[ImageInput], List[List[ImageInput]]] = None,
        text: Union[TextInput, "PreTokenizedInput", List[TextInput], List["PreTokenizedInput"]] = None,
        audio = ...
        videos: Union[str, List[ImageInput], List[List[ImageInput]]] = None,
        **kwargs: Unpack[...MyProcessorKwargs],
```

molbap commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,480 @@
+                `is_split_into_words=True` (to lift the ambiguity with a batch of sequences).
+                Wherever an image token, `<image>` is encountered it is expanded to
+                `<fake_token_around_image>` + `<row_x_col_y>` + `<image>` * `image_seq_len` * <fake_token_around_image>`.
+            image_seq_len (`int`, *optional*):
+                The length of the image sequence. If not provided, the default value of self.image_seq_len is used.
+                image_seq_len should be equal to int(((image_size // patch_size) ** 2) / (scale_factor**2))
```
I think this could always be inferred from config/model state without side effects, no?

orrzohar replied:
This was following Idefics3 processor logic; I removed it in smolVLM (I also think it is redundant, as I don't see why one would want to change this at inference time, it would break the model). For max_frames + fps -- this can be changed at inferenced, so i added it to the Kwargs class as suggested.

molbap commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,480 @@
+            **kwargs,
+        )
+
+        image_seq_len = image_seq_len if image_seq_len is not None else self.image_seq_len
+        video_sampling_fps = video_sampling_fps if video_sampling_fps is not None else self.video_sampling_fps
+        max_frames = max_frames if max_frames is not None else self.max_frames
```
these three extra kwargs should be moved to their respective dictionaries in `output_kwargs`

orrzohar replied:
done

molbap commented on src/transformers/models/smolvlm/video_processing_smolvlm.py:
```diff
@@ -0,0 +1,255 @@
+
+    # Centered skip if we want fewer frames than max_frames
+    if desired_frames < max_frames:
+        leftover = total_frames - desired_frames
+        start_idx = leftover // 2
+        end_idx = total_frames - (leftover - start_idx)
+    # Otherwise, if video is long enough, skip a bit from start and end
+    elif skip_secs > 0 and (duration_seconds - 2 * skip_secs) > (max_frames * target_fps):
+        start_idx = int(skip_secs * native_fps)
+        end_idx = int(total_frames - skip_secs * native_fps)
+
+    # Ensure valid start / end
+    start_idx = max(start_idx, 0)
+    end_idx = min(end_idx, total_frames - 1)
+    if start_idx >= end_idx:
+        start_idx = 0
+        end_idx = total_frames - 1
+
+    # Uniformly sample the desired number of frames from [start_idx..end_idx]
+    frames_idx = np.linspace(start_idx, end_idx, desired_frames, dtype=int)
+    frames_idx = np.unique(frames_idx).tolist()
+
+    # Read frames from decord
+    try:
+        frames_tensor = vr.get_batch(frames_idx).cpu().numpy()  # (N, H, W, C)
+    except Exception as e:
+        raise RuntimeError(f"Failed to read frames from '{path}': {e}")
+
+    # Convert to PIL Images
+    frames_out = [Image.fromarray(arr).convert("RGB") for arr in frames_tensor]
+
+    # Build timestamps (MM:SS) for each selected frame index
+    timestamps = []
+    for idx in frames_idx:
+        sec = idx / native_fps
+        mm = int(sec // 60)
+        ss = int(sec % 60)
+        timestamps.append(f"{mm:02d}:{ss:02d}")
+
+    return frames_out, timestamps, duration_seconds
```
As pointed out offline by @zucchini-nlp , the method `load_video` in `image_utils.py` is more generic and could be readily used, from what I understand you're skipping beginning & end frames optionally, I'd move this to a tinier postprocessing method and rely on `load_video`!

orrzohar replied:
We can't do this; we need the timestamps of the sampled frames.
Additionally, it does not support 'fps then uniform', where we sample a video at a set FPS until we reach max_frames, above which we start uniform frame sampling. Meanwhile, load_video:

https://github.com/huggingface/transformers/blob/11afab19c0e4b652855f9ed7f82aa010c4f14754/src/transformers/image_utils.py#L747

zucchini-nlp replied:
hey @orrzohar !

I skimmed through the loading function here. Yeah, it adds some stuff on top of the existing helper, but I think it will be much better to modify the existing one and move everything to `image_utils.py`. That will make code cleaner and easier for us to maintain, since we already have logic to load using different decoders and from url/path/youtube

We can decompose a bit the function, for example there already existsi `get_uniform_sampled_frames` and we can add extra kwargs like `start_idx` and `end_idx` to it. Currently it assumes start id is always `0`. In similar way, the "Centered skip if we want fewer frames than max_frames" part can be separated out, and called when needed. Maybe a boolean argument for that will be used, like `centered_skip=True`. 

The metadata as additional return is oke for me to add, we can return timestamps and durations in metadata dict. Later we can expand the dict if new techniques are used for lloading

orrzohar replied:
Hi @zucchini-nlp,

The main issue is that we do not just do uniform or fps sampling, we do 'fps then uniform' sampling, i.e., lets say we input fps=1 and max_frames=60, then for all videos shorter then 60 seconds we will do uniform frame sampling and all videos longer then 60 seconds we will do unform frame sampling of 60 frames from the entire video.

the skip seconds is not that important and can be dropped if we want.

zucchini-nlp replied:
sorry, not sure I got you. From what I see the model needs special way to obtain `desired_frames` is either a preset `num_frames` or obtained with `fps`. Currently we don't allow users to indicate both `num_frames` and `fps`, but I think we can enable and if both args are present we'll use the proposed idea

orrzohar replied:
@zucchini-nlp 
The idea is that we have a target video sampling FPS, but also our model cannot handle more then N frames -- so we want to sample at that FPS untill the video gets too long, at which point we switch to uniform frame sampling. 

for example, for fps=1, max_frames=60:
video duration = 30: we sample 30 frames at 1 fps
video duration = 70: we sample 60 frames uniformly from the video (6/7 effective FPS)

if load_video could do that + output metadata (frame timestamps in this case) then we could use this function instead of what i wrote

zucchini-nlp replied:
Yes, exactly! I believe we can do that, if we just move out the part with special sampling to a small function and call it within each `read_video_xxx`. LMK if you need help with writing that up 🤗

orrzohar replied:
@zucchini-nlp 

I think i have a better idea. Maybe you can add a 'get_video_details' function that gives video duration, num_frames, and fps?

If i can call this function to get these details, then calculate the indices I want to sample, we could instead pass to load_video the frame indicies we want to sample. This will allow more flexibility in the future (e.g., if future works want to use clip sampling, or other types of frame sampling)

zucchini-nlp replied:
yeah, sounds good! As long as all the logic is held inside utilities and can be later used up by other models, it works for me

orrzohar replied:
@zucchini-nlp 
why did you not use`get_uniform_frame_indices` in `read_video_torchvision` in accordance with all the other video loaders logic?

orrzohar replied:
I have updated load_video and added `get_video_details` and now we use load_video

zucchini-nlp replied:
Thanks! For torch it was only to keep the indices in torch tensor, but for consistency delegating sampling to a helper is a good idea. Given that now we sample frame ids in different ways

molbap commented on src/transformers/models/smolvlm/video_processing_smolvlm.py:
```diff
@@ -0,0 +1,255 @@
+    Returns: (frames, timestamps, duration_seconds).
+    This can be replaced by a more robust version with decord or ffmpeg, etc.
+    """
+    if is_url(path_or_url):
+        ## load video
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            filename = os.path.join(tmp_dir, "temp_video.mp4")
+            
+            # e.g. use requests
+            resp = requests.get(path_or_url, stream=True)
+            resp.raise_for_status()
+            with open(filename, "wb") as f:
+                for chunk in resp.iter_content(chunk_size=8192):
+                    f.write(chunk)
+            
+            # 2) Actually load frames from the local temp file
+            if is_decord_available():
+                frames, timestamps, duration_seconds = load_video(
+                    filename, max_frames=max_frames, target_fps=sampling_fps
+                )
+            elif is_cv2_available:
+                frames, timestamps, duration_seconds = load_video_cv2(
+                    filename, max_frames=max_frames, target_fps=sampling_fps
+                )
+            else:
+                raise RuntimeError("No valid video runtime. please install either opencv or decord")
+    
+    else:
+        if is_decord_available():
+            frames, timestamps, duration_seconds = load_video(
+                path_or_url, max_frames=max_frames, target_fps=sampling_fps
+            )
+        elif is_cv2_available:
+            frames, timestamps, duration_seconds = load_video_cv2(
+                path_or_url, max_frames=max_frames, target_fps=sampling_fps
+            )
+        else:
+            raise RuntimeError("No valid video runtime. please install either opencv or decord")
+            
+    return frames, timestamps, duration_seconds
```
Duplicate of load_video?

orrzohar replied:
we can't use load_video (see https://github.com/huggingface/transformers/pull/36126#discussion_r1951490770).

This allows us to either download video from url or load local video file

orrzohar replied:
I have updated load_video and added `get_video_details` and now we use load_video

molbap commented on src/transformers/models/smolvlm/video_processing_smolvlm.py:
```diff
@@ -0,0 +1,255 @@
+
+    start_idx = 0
+    end_idx = total_frames - 1
+
+    # If we want fewer than max_frames, center the skip
+    if desired_frames < max_frames:
+        leftover = total_frames - desired_frames
+        start_idx = leftover // 2
+        end_idx = total_frames - (leftover - start_idx)
+    # Otherwise skip some from start/end if feasible
+    elif skip_secs > 0 and (duration_seconds - 2 * skip_secs) > (max_frames * target_fps):
+        start_idx = int(skip_secs * native_fps)
+        end_idx = int(total_frames - skip_secs * native_fps)
+
+    start_idx = max(0, start_idx)
+    end_idx = min(total_frames - 1, end_idx)
+    if start_idx >= end_idx:
+        start_idx, end_idx = 0, total_frames - 1
+
+    frames_idx = np.linspace(start_idx, end_idx, desired_frames, dtype=int)
+    frames_idx = np.unique(frames_idx).tolist()
+
+    frames_out = []
+    timestamps = []
+
+    for idx in frames_idx:
+        cap.set(cv2.CAP_PROP_POS_FRAMES, idx)
+        success, frame = cap.read()
+        if not success or frame is None:
+            continue
+        frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
+        frames_out.append(Image.fromarray(frame_rgb).convert("RGB"))
+
+        sec = idx / native_fps
+        mm = int(sec // 60)
+        ss = int(sec % 60)
+        timestamps.append(f"{mm:02d}:{ss:02d}")
+
+    cap.release()
+    return frames_out, timestamps, duration_seconds
```
If we can import load_video, this should not be needed

orrzohar replied:
we can't use load_video (see https://github.com/huggingface/transformers/pull/36126#discussion_r1951490770).

molbap reviewed (commented):
Hey, I added a couple comments! LMK if something is unclear and ping me when you're done with these, I'll re-iterate quickly :hugs: 
And a first comment, make sure to run the formatter/linter/checker locally:
You'll need to install dev tools within transformers repo
```bash
pip install -e .[quality]
```
And then run this command, that'll cover all the checks to make CI happy
```bash
make fixup
```

molbap commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,480 @@
+            new_content = []
+            for block in msg["content"]:
+                if block.get("type") == "video":
+                    assert  num_frames is not None or timestamps is not None or duration_sec is not None,  "to use 'video' tokens, you must specify `num_frames`, `timestamps`, and `duration_sec`."
```
try/except rather than assert, in general

orrzohar replied:
done

molbap commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,480 @@
+            for block in msg["content"]:
+                if block.get("type") == "video":
+                    assert  num_frames is not None or timestamps is not None or duration_sec is not None,  "to use 'video' tokens, you must specify `num_frames`, `timestamps`, and `duration_sec`."
+                    # 1) Insert the intro
+                    # frames, timestamps, duration_sec must be provided
+                    # (Alternatively, you could dynamically load them here.)
+                    if num_frames is None or timestamps is None or duration_sec is None:
+                        # If user didn't pass these, raise or skip
+                        raise ValueError("Must provide num_frames, timestamps, and duration_sec to insert 'video' blocks.")
+
+                    # Build the video intro texts
+                    td = timedelta(seconds=duration_sec)
+                    new_content.append({"type": "text", "text": DEFAULT_VIDEO_INTRO.format(frame_count=num2words(num_frames), video_duration=str(td))})
+
+                    # 2) Insert per-frame lines: "Frame from {timestamp}:", then an "image" block
+                    for i, ts in enumerate(timestamps):
+                        new_content.append({"type": "text", "text": FRAME_TIMESTAMP_MESSAGE.format(timestamp=ts)})
+                        new_content.append({"type": "image"})
+
+                    # 3) Optionally add an outro (e.g. "Now answer the question:")
+                    new_content.append({"type": "text", "text": DEFAULT_MEDIA_OUTTRO})
+                    # Do NOT add the original block => we skip it (since we've replaced it)
+                else:
+                    # keep original block
+                    new_content.append(block)
```
Since chat templates can do jinja magic - it would be good to transfer as much as possible from the logic here to the `chat_template.json` file. Since `apply_chat_template` is done immediately I think it'd be better.

Ideally I'd prefer to not overload this method at all and rely on chat template, because you can pass parameters too there, but you need these external parameters coming from your `load_video` IIUC. 

Ideally: you can still get these parameters from a preprocessing method after `load_video`, then pass them to an overload mehtod that is smaller than the current one, and have the if/else insert logic directly baked into the chat template. Does that sound good?

molbap reviewed (commented):
Added a couple comments for `chat_template`, let's go!

molbap commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,480 @@
+            text_inputs = self.tokenizer(text=prompt_strings, **output_kwargs["text_kwargs"])
+            inputs.update(text_inputs)
+
+    def apply_chat_template(self, messages, add_generation_prompt, num_frames=None, timestamps=None, duration_sec=None):
```
Last comment on this method: overall I think if you can refactor so that the processor does this insertion logic work instead of the chat template, it'd be better and more maintainable as the current setup can create hard to debug model inputs!

orrzohar replied:
I have updated load_video and added `get_video_details` and now we use load_video

orrzohar commented on src/transformers/models/smolvlm/configuration_smolvlm.py:
```diff
@@ -0,0 +1,194 @@
+        layer_norm_eps=1e-6,
+        attention_dropout=0.0,
+        initializer_range=0.02,
+        max_frames=64,
```
Correct, I will add sampling_fps as well, but otherwise, this is the same. I can instead import the Idefics3 config as a superclass and add it if that is better.

orrzohar committed:
trying modular, but fails

orrzohar commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,480 @@
+            inputs.update(text_inputs)
+
+    def apply_chat_template(self, messages, add_generation_prompt, num_frames=None, timestamps=None, duration_sec=None):
+        """
+        Overrides apply_chat_template to first convert any {'type': 'video'} blocks
+        into a series of text+image references (video intro, frame placeholders, etc.).
+        Then calls the base class apply_chat_template.
+
+        If you already have frames/timestamps/duration, pass them in here, e.g. for
+        a single video scenario. In a more general or multi-video scenario, you might
+        expand this method or pass multiple sets of frames.
+
+        This method modifies 'messages' in-place.
+        """
```
I can instead support this by seperating the logic. 

I added a `process_video_token` to process the video token with the expected format (adding timestamps, video duration, etc.). Then, the superclass's apply_chat_template will work.

orrzohar committed:
modular is working, changing processor to match PR comments and general transformers logic

zucchini-nlp commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,480 @@
+        # or <row_x_col_y> tags (where x and y are digits, also optionally surrounded by newline characters).
+        self._regex_to_remove_extra_special_tokens = re.compile(r"(\n?<global-img>\n?|<row_\d+_col_\d+>\n?)+")
+
+        tokens_to_add = {
+            "additional_special_tokens": [
+                self.fake_image_token,
+                self.image_token,
+                self.end_of_utterance_token,
+            ]
+        }
+        tokenizer.add_special_tokens(tokens_to_add)
```
btw, we can now add these tokens to the tokenizer before release, and have it in config. See https://huggingface.co/docs/transformers/en/main_classes/tokenizer#multimodal-tokenizer :)

I think it will be easier that way to access special image tokens

orrzohar replied:
BTW this is legacy from idefics3 processor, we did not introduce this logic here

orrzohar replied:
@zucchini-nlp see:

https://github.com/huggingface/transformers/blob/11afab19c0e4b652855f9ed7f82aa010c4f14754/src/transformers/models/idefics3/processing_idefics3.py#L152-L159

I think we could potentially remove this, but not sure if will hurt backwards compatibility

zucchini-nlp replied:
yeah, just a suggestion since we have updated tokenizers recently and this is a new model checkpoint. We'll try to move away gradually from adding new tokens while initializing

orrzohar replied:
No problem then i'll remove this part

orrzohar replied:
@zucchini-nlp 
Correct me if I am wrong, but couldn't we then also clean `_regex_to_remove_extra_special_tokens` and `batch_decode`/`decode`? just rely on the new tokenizers?

zucchini-nlp replied:
if we want to remove all special tokens, then yes. Decoding only with `skip_special_tokens` should work without any regex sub, as long as all the tokens are added in special tokens map

orrzohar committed:
fixing kwargs

orrzohar commented:
@molbap @zucchini-nlp 
Overall comments:
1. I have committed a revision that uses modular transformers as suggested by @molbap 
2. I have refactored the inputs to the processor to be closer to the standard
3. I have removed our overwrite of apply_chat_template in favor of a new function that converts the video tokens into the expected sequence of text and image tokens, so we can just use the original apply_chat_template

orrzohar committed:
offloading video loading logic to  image_util

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

orrzohar committed:
fixing circleci code formatting errors

zucchini-nlp commented on src/transformers/models/smolvlm/video_processing_smolvlm.py:
```diff
@@ -0,0 +1,121 @@
+}
+
+
+def load_smolvlm_video(
+    path: str,
+    max_frames: int = 64,
+    sampling_fps: float = 1.0,
+    skip_secs: float = 1.0,
+    backend: str = "decord",
+) -> Tuple[List[Image.Image], List[str], float]:
+    """
+    Loads a video from `path` by first gathering metadata via `get_video_details`
```
I would aim at removing this entirely and moving logic to `image_utils`. Given that now we have another model that crops video from the middle, we need to have a general utility


So what I have in mind is:
- `load_video` will take extra kwargs (initial_shift, skip_secs)
- each decoder will receive these kwargs and infer how many frames to sample. Can be preset `num_frames` or inferred from `fps * duration`
- infer start and end indices using the below dummy logic
- sample indices with `np.linspace(start, end, num_frames)`
- sample frames and return metadata as a dict 

```python
def get_start_end_indices(initial_shift: Union[bool, int], skip_secs: Optional[float] = 0.0, metadata: Dict[str, any]):
    """
    Args:
        initial_shift (`bool`, `float` or `int`, defaults to `0`):
            The initial shift to apply when sampling frames. If `True`, the shift is set so that frames are sampled from the middle of the video.
        skip_secs (`float` or `int`, *optional*m defaults to `0.0`):
            Applied only when `initial_shift=True`. If cropping middle part of video results in a video longer than `max_frames` that can be sampled
            at given `fps`, then start/end indices for sampling are cropped to skip `skip_secs` seconds
    """
    duration, fps, etc = metadata.get(...)
    # Used by InternVL in certain cases
    if isinstance(initial_shift, (int, float)):
        start_idx = int(initial_shift)
        end_idx = total_num_frames
    # Copy from SmolVLM and can be used by InternVL (only when skip_secs==0)
    elif initial_shift:
        if skip_secs==0 or desired_frames < max_frames:
            leftover = total_num_frames - desired_frames
            start_idx = leftover // 2
            end_idx = total_num_frames - (leftover - start_idx)
        elif skip_secs > 0 and (duration_seconds - 2 * skip_secs) > (max_frames * sampling_fps):
            start_idx = int(skip_secs * fps)
            end_idx = int(total_num_frames - skip_secs * fps)
    # Old logic of uniform sampling from start to end
    else:
        start_idx = 0
        end_idx = total_num_frames
    return start_idx, end_idx
```

molbap replied:
Very nice - pro moving this to another PR either now or after release. Also an `Enum` would be cleaner and more easily extendable to future video processing utils

orrzohar committed:
fixing circleci code formatting errors

zucchini-nlp committed:
update

molbap committed:
add idefics3-based tests

molbap committed:
add keyword to all

molbap committed:
add PreTrainedModel

orrzohar committed:
Merge pull request #1 from zucchini-nlp/smolvlm2

updated  video loader + processor for SmolVLM2

orrzohar committed:
Merge pull request #2 from molbap/add_draft_test

Add draft test to smolvlm

orrzohar committed:
updateing video loading logic

molbap commented on src/transformers/models/smolvlm/modular_smolvlm.py:
```diff
@@ -0,0 +1,333 @@
+                    "Cannot map them to blocks of shape (S, D)."
+                )
+
+            # We'll chunk image_positions into groups of size S
+            positions_list = image_positions.tolist()
+            # Example: if num_image_tokens=162 and S=81 => we have 2 images => 2 chunks each of length 81
+            chunks = [positions_list[i : i + S] for i in range(0, num_image_tokens, S)]
+
+            # We'll build a list of segments: text, then image row(s), text, etc.
+            segments = []
+            text_start = 0
+
+            # For each chunk (each chunk => 1 image)
+            for chunk in chunks:
+                # image_hidden_states[image_offset] => shape (S, D)
+                cur_block = image_hidden_states[image_offset]
+                image_offset += 1
+
+                # We'll iterate over the S positions in ascending order
+                for i_s, pos in enumerate(chunk):
+                    # Add text from [text_start..pos)
+                    if pos > text_start:
+                        segments.append(cur_embeds[text_start:pos])
+                    # Then add one row from cur_block => shape (1, D)
+                    row_of_block = cur_block[i_s : i_s + 1, :]
+                    segments.append(row_of_block)
+                    # skip the <image> token
+                    text_start = pos + 1
+
+            # leftover text after the final <image> token
+            if text_start < T:
+                segments.append(cur_embeds[text_start:])
+
+            # cat them into a single (T_b, D) tensor
+            merged_sample = torch.cat(segments, dim=0)
+            merged_outputs.append(merged_sample)
+
+        merged_outputs = torch.stack(merged_outputs)
+        # assert (old_merger_outputs==merged_outputs).all()
+        return merged_outputs
```
I'd be pro-removing this and replacing it with a vectorized merger, I think a prefix sum can keep track of where the images are - I drafted something, I'm not sure it matches all cases but let's see

```suggestion
    batch_size, seq_length, hidden_size_text = inputs_embeds.shape
    num_image_blocks, patch_size, hidden_size_img = image_hidden_states.shape

    image_mask = (input_ids == image_token_id)
    num_image_tokens = image_mask.sum(dim=1)

    blocks_per_sample = num_image_tokens // patch_size
    
    offsets = F.pad(blocks_per_sample.cumsum(dim=0), (1, 0), value=0)
    block_offset = offsets[:-1]
    # I think something like this should work and separate chunks cleanly
    # TODO verify but for sure there's a low-hanging vectoriztion fruit
    row_cum = image_mask.cumsum(dim=-1)
    chunk_idx = (row_cum - 1) // patch_size
    local_idx = (row_cum - 1) % patch_size
    block_idx = block_offset.unsqueeze(1) + chunk_idx

    merged_outputs = inputs_embeds.clone()
    merged_outputs[image_mask] = image_hidden_states[block_idx[image_mask], local_idx[image_mask], :]

```

orrzohar replied:
This is not deepspeed-compatible, which is why we altered the original inputs merger

molbap commented on src/transformers/image_utils.py:
```diff
@@ -748,13 +802,29 @@ def load_video(
+
+    # If user didn't pass a sampling function, create one on the fly with default logic
+    if sample_indices_fn is None:
+        sample_indices_fn = lambda metadata, **fn_kwargs: default_sample_indices_fn(
```
this'll make CI yelp :P lambda functions are unwelcome

orrzohar replied:
done

molbap commented on src/transformers/models/smolvlm/modular_smolvlm.py:
```diff
@@ -0,0 +1,333 @@
+    """
+
+    def __init__(self, config):
+        super().__init__(config)
+        # Instead of the original self.model = Idefics3Model(config),
+        # we point to our custom class.
+        self.model = SmolVLMModel(config)
+
+        # We *keep* the same lm_head from the parent, or re-init if you prefer:
+        self.lm_head = nn.Linear(config.text_config.hidden_size, config.text_config.vocab_size, bias=False)
+
+        # If parent sets up any post_init() logic:
+        self.post_init()
```
Let's clean up inline comments

molbap commented on smolvlm.py:
to delete :)

molbap reviewed (commented):
Looking good! I suggest iterating on whatever the CI says and we can polish tomorrow

orrzohar committed:
working inference

orrzohar commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,466 @@
+            "return_row_col_info": True,
+        },
+        "videos_kwargs": {
+            "max_frames": None,
+            "sampling_fps": None,
+            "video_duration": None,
+        },
```
no the preprocessor_config.json will have these

orrzohar commented on tests/models/smolvlm/test_processor_smolvlm.py:
```diff
@@ -0,0 +1,578 @@
+        with self.assertRaises(ValueError):
+            processor(text=text, images=images, padding=True)
+
+    def test_apply_chat_template(self):
```
Can we fetch the video from the URL? otherwise, we can make mock video using np or similar.

zucchini-nlp replied:
Yeah, we definitely need tests for that. I will add some here and some more in general processing, for new chat template logic

My idea was to split out all video loading to new PR, lmk if that's what you prefer

orrzohar committed:
updates for PR comments

orrzohar committed:
updates for PR comments

orrzohar commented on src/transformers/models/smolvlm/modular_smolvlm.py:
```diff
@@ -0,0 +1,333 @@
+        B, T, D_text = inputs_embeds.shape
+        N, S, D_img = image_hidden_states.shape
+        if D_text != D_img:
+            raise ValueError(f"Text embedding dim {D_text} != image embedding dim {D_img}")
```
done

orrzohar committed:
moving SmolVLMPretrainedModel higher to fix import error

orrzohar committed:
CI test pass

orrzohar committed:
CI test pass

orrzohar committed:
removing lambda

orrzohar committed:
CI test pass

orrzohar committed:
CI test pass

orrzohar committed:
Merge branch 'main' into smolvlm

orrzohar committed:
CI test pass

orrzohar committed:
CI test pass

orrzohar committed:
CI test pass

orrzohar committed:
CI test pass

zucchini-nlp committed:
processor tests

zucchini-nlp committed:
add example in docs

zucchini-nlp committed:
typo

zucchini-nlp committed:
fix copies

molbap committed:
skip compile tests - sdpa for VisionTransformer

molbap committed:
fix init

molbap committed:
raise import error for num2words

molbap committed:
update doc for FA2

molbap committed:
more doc fix

molbap committed:
Merge branch 'smolvlm' of github.com:orrzohar/transformers into smolvlm

molbap committed:
CI

pcuenca reviewed (commented):
🔥

pcuenca commented on docs/source/en/model_doc/smolvlm.md:
```diff
@@ -0,0 +1,197 @@
+
+Videos should not be upsampled. 
+
+If `do_resize` is set to `True`, the model resizes images so that the longest edge is 4*364 pixels by default.
+The default resizing behavior can be customized by passing a dictionary to the `size` parameter. For example, `{"longest_edge": 4 * 364}` is the default, but you can change it to a different value if needed.
```
```suggestion
If `do_resize` is set to `True`, the model resizes images so that the longest edge is 4*384 pixels by default.
The default resizing behavior can be customized by passing a dictionary to the `size` parameter. For example, `{"longest_edge": 4 * 384}` is the default, but you can change it to a different value if needed.
```

I think, cc @mfarre

orrzohar replied:
We actually have both, depending on the model

pcuenca replied:
But what is the default?

orrzohar replied:
The default should be from the processor_config.json, should not be hard-encoded in code - right @zucchini-nlp ?

zucchini-nlp replied:
right

pcuenca commented on docs/source/en/model_doc/smolvlm.md:
````diff
@@ -0,0 +1,197 @@
+image_processor = SmolVLMImageProcessor(do_resize=True, size={"longest_edge": 2 * 364}, max_image_size=364)
+```
+
+Additionally, the `max_image_size` parameter, which controls the size of each square patch the image is decomposed into, is set to 364 by default but can be adjusted as needed. After resizing (if applicable), the image processor decomposes the images into square patches based on the `max_image_size` parameter.
````
```suggestion
Additionally, the `max_image_size` parameter, which controls the size of each square patch the image is decomposed into, is set to 384 by default but can be adjusted as needed. After resizing (if applicable), the image processor decomposes the images into square patches based on the `max_image_size` parameter.
```

orrzohar replied:
The model can handle both image and video inputs, and we do image splitting for image/multi-image inference/training. I'll see if i can hard-encode this in the processor.

pcuenca replied:
This was just meant as a comment on the docs.

pcuenca commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,443 @@
+
+        sample_indices_fn = sample_indices_fn_func
+        return super().apply_chat_template(
+            conversation, sample_indices_fn=sample_indices_fn, video_load_backend="decord", **kwargs
```
This implies we depend on `decord`. Is this intended?

zucchini-nlp replied:
@orrzohar can we remove it? Let's allow users to set the backend + `decord` is not very stable. We had many issues in the past with decord (https://github.com/huggingface/transformers/issues/21085), thus we have `pyav` as default value

zucchini-nlp commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -797,6 +800,7 @@
         ("pix2struct", "Pix2StructForConditionalGeneration"),
         ("qwen2_5_vl", "Qwen2_5_VLForConditionalGeneration"),
         ("qwen2_vl", "Qwen2VLForConditionalGeneration"),
+        ("smolvlm", "SmolVLMForConditionalGeneration"),
```
let's remove this before merging, auto-map should be only for ImageTextToText in VLMs

pcuenca commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,443 @@
+        self.end_of_utterance_token = tokenizer.end_of_utterance_token
+        self.global_image_token = tokenizer.global_image_token  # https://github.com/huggingface/transformers/pull/32473/files/8063e5e17362571b693f1db95167f5443a3be1b2#r1734825341
+        self.image_seq_len = image_seq_len
+        self.video_frame_size = image_processor.video_sampling["video_size"]
```
This defaults to `384` in my tests, whereas image_size is `512`. This, in turn, results in the frames being scaled to `384` first, and then to `512`.

orrzohar replied:
Our 2b model accepts 384 images, while our smaller ones accept 512.
@molbap / @zucchini-nlp Any suggestions? where is this being defined?

I think the value should be pulled from our json configs, not hard-encoded and then this would not be an issue

orrzohar committed:
updates for PR comments

orrzohar committed:
Update docs/source/en/model_doc/smolvlm.md

Co-authored-by: Pedro Cuenca <pedro@huggingface.co>

orrzohar committed:
Update docs/source/en/model_doc/smolvlm.md

Co-authored-by: Pedro Cuenca <pedro@huggingface.co>

orrzohar committed:
Update docs/source/en/model_doc/smolvlm.md

Co-authored-by: Joshua Lochner <admin@xenova.com>

orrzohar committed:
Update docs/source/en/model_doc/smolvlm.md

Co-authored-by: Pedro Cuenca <pedro@huggingface.co>

orrzohar committed:
Update docs/source/en/model_doc/smolvlm.md

Co-authored-by: Pedro Cuenca <pedro@huggingface.co>

orrzohar committed:
fixing processor -- tokenizer not defined properly, (gpt2 tokenizer), and does not have the attributes of fake image token, etc

xenova commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,459 @@
+                per batch.
+        """
+        # We don't want to modify in-place the messages passed by user
+        # The user might want to add new turn on conv and continue geenration
```
```suggestion
        # The user might want to add new turn on conv and continue generation
```

orrzohar committed:
adding smolvlm to VQA models

orrzohar committed:
removing vqa auto class

orrzohar committed:
Update src/transformers/models/smolvlm/processing_smolvlm.py

Co-authored-by: Joshua Lochner <admin@xenova.com>

orrzohar committed:
updating from upstream for PR

orrzohar committed:
removing smolvlmvisiontransformer from index.md

zucchini-nlp commented on src/transformers/models/smolvlm/modular_smolvlm.py:
```diff
@@ -0,0 +1,295 @@
+        image_offset = 0
+        merged_outputs: List[torch.Tensor] = []
+
+        # Iterate through each sample
+        for b_idx, (cur_ids, cur_embeds) in enumerate(zip(input_ids, inputs_embeds)):
+            # Find positions of <image> tokens in the text
+            image_positions = (cur_ids == self.image_token_id).nonzero(as_tuple=True)[0]
+            num_image_tokens = len(image_positions)
+
+            # If no <image> => text-only
+            if num_image_tokens == 0:
+                # NOTE: this is important for DeepSpeed.
+                empty_slice = image_hidden_states[0][:0, :]  # shape (0, D)
+                merged_text_only = torch.cat([cur_embeds, empty_slice], dim=0)
+                merged_outputs.append(merged_text_only)
+                continue
+
+            # Typically, if each image is S embeddings, we expect the total # of <image> tokens
+            # in this sample to be multiple of S => each group of S tokens = 1 image
+            if num_image_tokens % S != 0:
+                raise ValueError(
+                    f"Sample {b_idx} has {num_image_tokens} <image> tokens, not a multiple of S={S}. "
+                    "Cannot map them to blocks of shape (S, D)."
+                )
+
+            positions_list = image_positions.tolist()
+            chunks = [positions_list[i : i + S] for i in range(0, num_image_tokens, S)]
```
I remember this was discusssed with Pablo previously. Now when I look at it, I don;t get why it is important to loop over each image position for Deepspeed. In qwen2-vl we do simple `masked_scatter` and the model is deepspeed compatible. Can you explain if the below doesn't work and why, will save us number of iterations from core maintainer

Like this
```python
image_mask = (
                (input_ids == self.image_token_id)
                .unsqueeze(-1)
                .expand_as(inputs_embeds)
                .to(inputs_embeds.device)
            )
            image_hidden_states = image_hidden_states.to(inputs_embeds.device, inputs_embeds.dtype)
            inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_hidden_states)
            return inputs_embeds
```

orrzohar replied:
When training on data mixtures (text/image/video), the forward pass (no matter the modality) needs to be identical.
If you have a text-only batch, you still need to encode an image and concatenate a (0-dim) tensor to the input embeddings.

If masked_scatter is deepspeed compatible (are you sure? I can find almost no support for this function), you still need to concat an empty image tensor for text-only batches.

zucchini-nlp replied:
Is that a training-only issue? 

I cannot say about docs, but Qwen2-VL is for sure working on DeepSpeed as per user's comments.

orrzohar replied:
only when training with zero3 or multi-node zero2, this has to do with FSDP needing exactly the same computational graph during grad accumulation across model copies -- if not the same, it doesn't know what to do and just hangs. (e.g., you can't train idefics2/idefics3 from experience with deepspeed).

zucchini-nlp replied:
oke, thanks, I'll leave this for Arthur to decide if we need to support it. Ping me on slack when comments are addressed and CI is green, will have one final look and request Arthur's review :)

zucchini-nlp committed:
my bad, video processing had typos

orrzohar committed:
fixing docs

orrzohar commented on src/transformers/models/smolvlm/configuration_smolvlm.py:
```diff
@@ -0,0 +1,197 @@
+    SmolVLM vision encoder according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the SigLIP checkpoint
+    [google/siglip-base-patch16-224](https://huggingface.co/google/siglip-base-patch16-224) used in the SmolVLM model
+    [HuggingFaceM4/SmolVLM-8B-Llama3](https://huggingface.co/HuggingFaceM4/SmolVLM-8B-Llama3).
```
done

orrzohar committed:
renaming params in SmolVLMModel.inputs_merger

orrzohar commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -372,6 +374,7 @@
         ("roberta-prelayernorm", "RobertaPreLayerNormForMaskedLM"),
         ("roc_bert", "RoCBertForPreTraining"),
         ("rwkv", "RwkvForCausalLM"),
+        ("smolvlm", "SmolVLMForConditionalGeneration"),
```
done

orrzohar commented on src/transformers/models/smolvlm/modular_smolvlm.py:
```diff
@@ -0,0 +1,295 @@
+        """
+
+        B, T, D_text = inputs_embeds.shape
+        N, S, D_img = image_hidden_states.shape
```
done

orrzohar committed:
removing un-needed dtype/device in model forward

orrzohar committed:
ruff for CI

zucchini-nlp commented on src/transformers/models/smolvlm/modular_smolvlm.py:
```diff
@@ -0,0 +1,427 @@
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
```
cache_position passed here also, to the tex model

orrzohar replied:
idk why this is not shown as outdated, but it is here now:

zucchini-nlp reviewed (approved):
Thanks!

orrzohar committed:
update docs

orrzohar committed:
Update docs/source/en/model_doc/smolvlm.md

Co-authored-by: Pedro Cuenca <pedro@huggingface.co>

orrzohar committed:
return cache position

orrzohar committed:
return cache position

orrzohar committed:
return cache also in modular

orrzohar committed:
needed to run modular again

ArthurZucker commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,461 @@
+    }
+
+
+SmolVLMProcessorKwargs.__annotations__["images_kwargs"] = SmolVLMImagesKwargs  # python 3.8 compatibility
```
we no longer support 3.8 so not needed

ArthurZucker commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,461 @@
+        the docstring of this method for more information.
+        """
+        decode_output = self.tokenizer.decode(*args, **kwargs)
+        return self._regex_to_remove_extra_special_tokens.sub("<image>", decode_output)
```
sorry not sure I undertsand this?

ArthurZucker commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,461 @@
+        refer to the docstring of this method for more information.
+        """
+        batched_decode_output = self.tokenizer.batch_decode(*args, **kwargs)
+        return [self._regex_to_remove_extra_special_tokens.sub("<image>", s) for s in batched_decode_output]
```
neither do I get why we have to do this?

ArthurZucker commented on src/transformers/models/smolvlm/processing_smolvlm.py:
```diff
@@ -0,0 +1,461 @@
+
+        self.default_max_frames = image_processor.video_sampling["max_frames"]
+        self.default_fps = image_processor.video_sampling["fps"]
+        # Matches one or more occurrences of <row_x_col_y> tags (where x and y are digits, optionally surrounded by newline characters
+        self._regex_to_remove_extra_special_tokens = re.compile(r"(<row_\d+_col_\d+>\n?)+")
```
if you add them as special tokens you won't go through all the trouble. These tokens are processed by the model anyways no?

orrzohar replied:
I left this to stay compatible with smolvlm1 -- I think we can drop it if we don't want to support that class here (I will just need to test).

ArthurZucker commented on src/transformers/models/smolvlm/modular_smolvlm.py:
```diff
@@ -0,0 +1,428 @@
+        elif image_hidden_states is not None:
+            image_hidden_states = image_hidden_states.to(dtype=self.dtype, device=input_ids.device)
+
+        if past_seen_tokens == 0 and inputs_embeds is not None and image_hidden_states is not None:
+            # When we generate, we don't want to replace the potential image_token_id that we generated by images
+            # that simply don't exist
+            inputs_embeds = self.inputs_merger(
+                input_ids=input_ids,
+                inputs_embeds=inputs_embeds,
+                image_hidden_states=image_hidden_states,
+            )
```
Question again here, but this disables the multi-turn image processing. Basically what if you want to use a different video/ pass other image?

ArthurZucker reviewed (approved):
Great work everyone! 
Not super fan of what's happening with the processor / video processor only doing half of the job each, we really need to split the work, keep it simpler: 
video processor just returns post processed videos / sampled frames and some metadata, while the processor should merge text and these! 

Main question: quid of multiturn

molbap committed:
fix training tests

molbap committed:
push vectorized inputs merger

molbap committed:
format

molbap committed:
format

molbap commented:
Pushed a fix for the test_training being unhappy + an attempt at vectorizing the merger that seems to work, would need another eye on it :)

zucchini-nlp commented:
@molbap thanks! I see you also added back the auto-map. The model was removed from `MODEL_FOR_VISION_2_SEQ_MAPPING_NAMES` intentionally. We should aim to use only the ImageTextToText mapping for VLMs, instead of duplicating over two auto-classes

molbap commented:
Yeah, it's just that if the model is not on the correct mappings, `test_training` fails unless we add manually some `labels` (else, they are not added by the auto class mapper). + Idefics3 had this double mapping AFAIK?
+1 to removing it though as long as tests pass, esp. training tests!

zucchini-nlp commented:
@molbap I think we can add the `ImageTextToText` mapping in `prepare_config_and_inputs_for_common` to make tests happy, along with CausalMap etc. Idefics3 indeed has both, and all other VLMs too, but it doesn't really make sense to use Vision2Seq for new models. Especially after we released the pipeline and have been promoting the `image-text-to-text` tag on the hub

molbap committed:
reduce number of mappings

orrzohar committed:
addressing PR comments

zucchini-nlp committed:
happy CI, happy me :)

molbap committed:
skip non-nested images

molbap committed:
adjust integration test for smaller GPUs

molbap committed:
format

molbap committed:
fix kwargs in chat template apply

molbap committed:
skip this for now

molbap committed:
Merge branch 'main' into smolvlm

LysandreJik commented:
Thanks all!
